test/integration/grpc/reconnect.tap.js

```diff
@@ -114,6 +115,93 @@ tap.test(
   }
 )
 
+/**
+ * With Node streams, when the server sends back data and we are not subscribed to the
+ * data event, the status event never fires thus preventing reconnects. This results in
+ * us being pinned to a bad stream and throwing 'ERR_STREAM_WRITE_AFTER_END' errors.
+ */
+tap.test(
+  'Should reconnect even when data sent back',
+  {skip:isUnsupportedNodeVersion},
+  async t => {
+    // one assert for the initial connection
+    // a second assert for the disconnect
+    // a third assert for the reconnection
+    // a fourth assert for the disconnect
+    // a fifth assert for server connection count
+    t.plan(5)
+
+    let serverConnections = 0
+
+    const recordSpan = (stream) => {
+      serverConnections++
+
+      // drain reads to make sure everything finishes properly
+      stream.on('data', () => {
+        // Writing data critical for triggering bug.
+        stream.write({messages_seen: 1})
+
+        // end the stream -- sends back a STATUS OK
+        stream.end()
+      })
+    }
+
+    const sslOpts = await setupSsl()
+    const port = await setupServer(t, sslOpts, recordSpan)
+
+    // Currently test-only configuration
+    const origEnv = process.env.NEWRELIC_GRPCCONNECTION_CA
+    process.env.NEWRELIC_GRPCCONNECTION_CA = sslOpts.ca
+    t.teardown(() => {
+      process.env.NEWRELIC_GRPCCONNECTION_CA = origEnv
+    })
+
+    return new Promise((resolve) => {
+      const metrics = createMetricAggregatorForTests()
+
+      const traceObserverConfig = {
+        host: 'ssl.lvh.me',
+        port: port
+      }
+
+      // very short backoff to trigger the reconnect in 1 second
+      const backoffs = {initialSeconds: 0, seconds:1}
+      const connection = new GrpcConnection(traceObserverConfig, metrics, backoffs)
+
+      let countDisconnects = 0
+
+      connection.on('connected', (callStream) => {
+        t.equals(
+          callStream.constructor.name,
+          'ClientDuplexStreamImpl',
+          'connected and received ClientDuplexStreamImpl'
+        )
+
+        callStream.write(new StreamingSpanEvent())
+      })
+
+      connection.on('disconnected', () => {
+        countDisconnects++
+        t.ok(true, 'disconnected')
+
+        // if we've disconnected twice, the test is done
+        // mark the state as permanantly closed in order to
+        // avoid further automatic reconnects (skipping
+        // _setState to avoid an additional disconnect event)
+        if (countDisconnects > 1) {
+          connection._state = 3 // replace with actual fake enum
```

Comment:
minor: maybe add TODO here? or even just comment here that 3 is permanently closed. Comments above are great so, no really urgency to update here.

Reply:
Ah yeah, copy-paste from Alan's original test case. Prob leave for now. Eventually, we can prob extract some of this to some shared logic anyways.